File: packages/grab_go_shared/lib/shared/services/credit_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'secure_storage_service.dart';

/// Credit transaction model
class CreditTransaction {
  final String id;
  final double amount;
  final String formattedAmount;
  final String type;
  final String typeLabel;
  final String? description;
  final String? orderId;
  final DateTime createdAt;
  final bool isCredit;

  CreditTransaction({
    required this.id,
    required this.amount,
    required this.formattedAmount,
    required this.type,
    required this.typeLabel,
    this.description,
    this.orderId,
    required this.createdAt,
    required this.isCredit,
  });

  factory CreditTransaction.fromJson(Map<String, dynamic> json) {
    return CreditTransaction(
      id: json['id'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      formattedAmount: json['formattedAmount'] ?? '',
      type: json['type'] ?? '',
      typeLabel: json['typeLabel'] ?? '',
      description: json['description'],
      orderId: json['orderId'],
      createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
      isCredit: json['isCredit'] ?? false,
    );
  }
}

/// Credit balance model
class CreditBalance {
  final double balance;
  final String currency;
  final String formatted;

  CreditBalance({required this.balance, required this.currency, required this.formatted});

  factory CreditBalance.fromJson(Map<String, dynamic> json) {
    return CreditBalance(
      balance: (json['balance'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'GHS',
      formatted: json['formatted'] ?? '₵0.00',
    );
  }
}

/// Credit calculation result for checkout
class CreditCalculation {
  final double creditsApplied;
  final double remainingPayment;
  final double creditBalance;
  final String formattedCreditsApplied;
  final String formattedRemainingPayment;
  final String formattedCreditBalance;

  CreditCalculation({
    required this.creditsApplied,
    required this.remainingPayment,
    required this.creditBalance,
    required this.formattedCreditsApplied,
    required this.formattedRemainingPayment,
    required this.formattedCreditBalance,
  });

  factory CreditCalculation.fromJson(Map<String, dynamic> json) {
    return CreditCalculation(
      creditsApplied: (json['creditsApplied'] ?? 0).toDouble(),
      remainingPayment: (json['remainingPayment'] ?? 0).toDouble(),
      creditBalance: (json['creditBalance'] ?? 0).toDouble(),
      formattedCreditsApplied: json['formattedCreditsApplied'] ?? '₵0.00',
      formattedRemainingPayment: json['formattedRemainingPayment'] ?? '₵0.00',
      formattedCreditBalance: json['formattedCreditBalance'] ?? '₵0.00',
    );
  }
}

/// GrabGo Credit Service
/// Manages in-app store credits for customers
class CreditService {
  static final CreditService _instance = CreditService._internal();
  factory CreditService() => _instance;
  CreditService._internal();

  static const String _baseUrl = 'https://grabgo-backend.onrender.com/api/credits';

  /// Get authorization headers
  Future<Map<String, String>> _getHeaders() async {
    final token = await SecureStorageService.getAuthToken();
    return {'Content-Type': 'application/json', 'Authorization': 'Bearer $token'};
  }

  /// Get current user's credit balance
  Future<CreditBalance?> getBalance() async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(Uri.parse('$_baseUrl/balance'), headers: headers);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success'] == true) {
          return CreditBalance.fromJson(data['data']);
        }
      }
      debugPrint('Failed to get credit balance: ${response.body}');
      return null;
    } catch (e) {
      debugPrint('Error getting credit balance: $e');
      return null;
    }
  }

  /// Get credit transaction history
  Future<List<CreditTransaction>> getTransactionHistory({int page = 1, int limit = 20}) async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(Uri.parse('$_baseUrl/transactions?page=$page&limit=$limit'), headers: headers);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success'] == true) {
          final transactions = (data['data']['transactions'] as List)
              .map((tx) => CreditTransaction.fromJson(tx))
              .toList();
          return transactions;
        }
      }
      debugPrint('Failed to get transactions: ${response.body}');
      return [];
    } catch (e) {
      debugPrint('Error getting transactions: $e');
      return [];
    }
  }

  /// Calculate credit application for checkout
  Future<CreditCalculation?> calculateForCheckout(double orderTotal, {bool useCredits = true}) async {
    try {
      final headers = await _getHeaders();
      final response = await http.post(
        Uri.parse('$_baseUrl/calculate'),
        headers: headers,
        body: jsonEncode({'orderTotal': orderTotal, 'useCredits': useCredits}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success'] == true) {
          return CreditCalculation.fromJson(data['data']);
        }
      }
      debugPrint('Failed to calculate credits: ${response.body}');
      return null;
    } catch (e) {
      debugPrint('Error calculating credits: $e');
      return null;
    }
  }

  /// Check if user has any credits available
  Future<bool> hasCredits() async {
    final balance = await getBalance();
    return balance != null && balance.balance > 0;
  }

  /// Get formatted balance string
  Future<String> getFormattedBalance() async {
    final balance = await getBalance();
    return balance?.formatted ?? '₵0.00';
  }
}

File: backend/services/credit_service.js
const prisma = require("../config/prisma");
const { sendToUser } = require("./fcm_service");
const { createNotification } = require("./notification_service");
const { getIO } = require("../utils/socket");

/**
 * GrabGo Credit Service
 * Manages in-app store credits for customers
 * Credits are non-withdrawable and non-transferable
 */
class CreditService {
  /**
   * Get user's current credit balance
   * @param {string} userId - User ID
   * @returns {Promise<number>} Credit balance
   */
  async getBalance(userId) {
    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: { creditBalance: true },
    });
    return user?.creditBalance || 0;
  }

  /**
   * Get user's credit transaction history
   * @param {string} userId - User ID
   * @param {object} options - Pagination options
   * @returns {Promise<object>} Transactions with pagination
   */
  async getTransactionHistory(userId, { page = 1, limit = 20 } = {}) {
    const skip = (page - 1) * limit;

    const [transactions, total] = await Promise.all([
      prisma.userCredit.findMany({
        where: { userId },
        orderBy: { createdAt: "desc" },
        skip,
        take: limit,
        select: {
          id: true,
          amount: true,
          type: true,
          description: true,
          orderId: true,
          createdAt: true,
        },
      }),
      prisma.userCredit.count({ where: { userId } }),
    ]);

    return {
      transactions,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  /**
   * Grant credits to a user (adds to balance)
   * @param {object} params - Grant parameters
   * @returns {Promise<object>} Transaction record
   */
  async grantCredits({
    userId,
    amount,
    type,
    description,
    orderId = null,
    referralId = null,
    adminId = null,
  }) {
    if (amount <= 0) {
      throw new Error("Credit amount must be positive");
    }

    // Use transaction to ensure atomicity
    const result = await prisma.$transaction(async (tx) => {
      // Create credit transaction record
      const creditTransaction = await tx.userCredit.create({
        data: {
          userId,
          amount,
          type,
          description,
          orderId,
          referralId,
          isUsed: false,
          isActive: true,
        },
      });

      // Update user's cached balance
      await tx.user.update({
        where: { id: userId },
        data: {
          creditBalance: { increment: amount },
        },
      });

      // Get updated balance
      const user = await tx.user.findUnique({
        where: { id: userId },
        select: { creditBalance: true },
      });

      return {
        transaction: creditTransaction,
        newBalance: user.creditBalance,
      };
    });

    console.log(
      `💰 [Credit] Granted ₵${amount.toFixed(2)} to user ${userId} (${type}). New balance: ₵${result.newBalance.toFixed(2)}`
    );

    // Send notification for credit received
    await this._sendCreditNotification(userId, amount, type, description, result.newBalance);

    return result;
  }

  /**
   * Send notification when credits are granted
   */
  async _sendCreditNotification(userId, amount, type, description, newBalance) {
    try {
      const typeLabels = {
        referral_earned: "Referral Reward",
        referral_received: "Referral Bonus",
        promotion: "Promotional Credit",
        refund: "Refund Credit",
        bonus: "Bonus Credit",
        admin_grant: "Credit Added",
      };

      const title = `💰 ${typeLabels[type] || "Credits Received"}`;
      const body = `₵${amount.toFixed(2)} has been added to your GrabGo Credits. New balance: ₵${newBalance.toFixed(2)}`;

      // Send FCM push notification
      await sendToUser(
        userId,
        { title, body },
        { type: "credit_received", amount: amount.toString(), newBalance: newBalance.toString() }
      );

      // Create in-app notification
      const io = getIO();
      if (io) {
        await createNotification(
          userId,
          "credit",
          title,
          description || body,
          { type: "credit_received", amount, newBalance, route: "/credits" },
          io
        );
      }
    } catch (error) {
      console.error("Error sending credit notification:", error.message);
    }
  }

  /**
   * Deduct credits from a user (subtracts from balance)
   * @param {object} params - Deduction parameters
   * @returns {Promise<object>} Transaction record
   */
  async deductCredits({
    userId,
    amount,
    type,
    description,
    orderId = null,
    adminId = null,
  }) {
    if (amount <= 0) {
      throw new Error("Deduction amount must be positive");
    }

    // Check current balance
    const currentBalance = await this.getBalance(userId);
    if (currentBalance < amount) {
      throw new Error(
        `Insufficient credit balance. Available: ₵${currentBalance.toFixed(2)}, Requested: ₵${amount.toFixed(2)}`
      );
    }

    // Use transaction to ensure atomicity
    const result = await prisma.$transaction(async (tx) => {
      // Create deduction transaction record (negative amount)
      const creditTransaction = await tx.userCredit.create({
        data: {
          userId,
          amount: -amount, // Negative for deductions
          type,
          description,
          orderId,
          isUsed: true,
          isActive: true,
          usedAt: new Date(),
        },
      });

      // Update user's cached balance
      await tx.user.update({
        where: { id: userId },
        data: {
          creditBalance: { decrement: amount },
        },
      });

      // Get updated balance
      const user = await tx.user.findUnique({
        where: { id: userId },
        select: { creditBalance: true },
      });

      return {
        transaction: creditTransaction,
        newBalance: user.creditBalance,
      };
    });

    console.log(
      `💸 [Credit] Deducted ₵${amount.toFixed(2)} from user ${userId} (${type}). New balance: ₵${result.newBalance.toFixed(2)}`
    );

    return result;
  }

  /**
   * Apply credits to an order at checkout
   * Returns the amount of credits to apply and remaining payment amount
   * @param {string} userId - User ID
   * @param {number} orderTotal - Total order amount
   * @param {boolean} useCredits - Whether user wants to use credits
   * @returns {Promise<object>} Credit application details
   */
  async calculateCreditApplication(userId, orderTotal, useCredits = true) {
    if (!useCredits) {
      return {
        creditsApplied: 0,
        remainingPayment: orderTotal,
        creditBalance: await this.getBalance(userId),
      };
    }

    const creditBalance = await this.getBalance(userId);
    const creditsToApply = Math.min(creditBalance, orderTotal);
    const remainingPayment = orderTotal - creditsToApply;

    return {
      creditsApplied: creditsToApply,
      remainingPayment: Math.max(0, remainingPayment),
      creditBalance,
    };
  }

  /**
   * Apply credits to an order (called after payment success or for credit-only orders)
   * @param {string} userId - User ID
   * @param {string} orderId - Order ID
   * @param {number} amount - Amount of credits to apply
   * @returns {Promise<object>} Transaction result
   */
  async applyCreditsToOrder(userId, orderId, amount) {
    if (amount <= 0) {
      return { creditsApplied: 0, newBalance: await this.getBalance(userId) };
    }

    return await this.deductCredits({
      userId,
      amount,
      type: "order_payment",
      description: `Credits applied to order`,
      orderId,
    });
  }

  /**
   * Refund order amount as credits
   * @param {string} userId - User ID
   * @param {string} orderId - Order ID
   * @param {number} amount - Refund amount
   * @param {string} reason - Refund reason
   * @returns {Promise<object>} Transaction result
   */
  async refundAsCredits(userId, orderId, amount, reason = "Order refund") {
    return await this.grantCredits({
      userId,
      amount,
      type: "refund",
      description: reason,
      orderId,
    });
  }

  /**
   * Admin grant credits to user
   * @param {object} params - Grant parameters
   * @returns {Promise<object>} Transaction result
   */
  async adminGrantCredits({ userId, amount, reason, adminId }) {
    return await this.grantCredits({
      userId,
      amount,
      type: "admin_grant",
      description: `Admin credit: ${reason}`,
      adminId,
    });
  }

  /**
   * Admin deduct credits from user
   * @param {object} params - Deduction parameters
   * @returns {Promise<object>} Transaction result
   */
  async adminDeductCredits({ userId, amount, reason, adminId }) {
    return await this.deductCredits({
      userId,
      amount,
      type: "admin_deduct",
      description: `Admin deduction: ${reason}`,
      adminId,
    });
  }

  /**
   * Grant promotional credits
   * @param {object} params - Promo parameters
   * @returns {Promise<object>} Transaction result
   */
  async grantPromoCredits({ userId, amount, promoName }) {
    return await this.grantCredits({
      userId,
      amount,
      type: "promotion",
      description: promoName || "Promotional credit",
    });
  }

  /**
   * Grant welcome credits for new users
   * @param {string} userId - User ID
   * @param {number} amount - Welcome credit amount (default ₵5)
   * @returns {Promise<object>} Transaction result
   */
  async grantWelcomeCredits(userId, amount = 5.0) {
    return await this.grantCredits({
      userId,
      amount,
      type: "bonus",
      description: "Welcome to GrabGo! 🎉",
    });
  }

  /**
   * Rollback credit usage if payment fails
   * @param {string} userId - User ID
   * @param {string} orderId - Order ID
   * @param {number} amount - Amount to restore
   * @returns {Promise<object>} Transaction result
   */
  async rollbackCreditUsage(userId, orderId, amount) {
    if (amount <= 0) return null;

    return await this.grantCredits({
      userId,
      amount,
      type: "refund",
      description: "Credit restored - payment failed",
      orderId,
    });
  }

  /**
   * Recalculate user's credit balance from transaction history
   * Used for data integrity checks
   * @param {string} userId - User ID
   * @returns {Promise<object>} Recalculation result
   */
  async recalculateBalance(userId) {
    const result = await prisma.userCredit.aggregate({
      where: { userId, isActive: true },
      _sum: { amount: true },
    });

    const calculatedBalance = result._sum.amount || 0;

    const user = await prisma.user.update({
      where: { id: userId },
      data: { creditBalance: calculatedBalance },
      select: { creditBalance: true },
    });

    console.log(
      `🔄 [Credit] Recalculated balance for user ${userId}: ₵${(calculatedBalance || 0).toFixed(2)}`
    );

    return {
      userId,
      newBalance: user.creditBalance,
    };
  }
}

module.exports = new CreditService();
